### PR TITLE
http-parser-fixes - intermediate review

### PR DESCRIPTION
**Pull request by vdmit11**

### CODE REVIEW

File: tempesta_fw/http.c [hunk 1]
```diff
@@ -118,8 +118,8 @@ tfw_http_establish_skb_hdrs(TfwHttpMsg *hm)
 		TfwHttpHdr *hdr = hm->h_tbl->tbl + i;
 		if (hdr->skb)
 			continue;
-		if (!hdr->field.ptr)
-			break;
+		if (!hdr->field.chunks)
+			break;  /* XXX: why do we stop here? */
```
Comment (krizhanovsky) [inline tempesta_fw/http.c]: Long time ago headers were placed one after other, so we had continous headers array and empty header meant end on the array. Not is not the case and the break must be fixed.

File: tempesta_fw/http_parser.c [hunk 2]
```diff
@@ -864,64 +874,47 @@ __parse_transfer_encoding(TfwHttpMsg *msg, unsigned char *data, size_t *lenrval)
...
+static TfwHttpHdr *
+__alloc_header(TfwHttpMsg *hm, tfw_http_hdr_t id)
 {
 	TfwHttpHdrTbl *ht = hm->h_tbl;
-	TfwStr *h;
 
-	if (unlikely(id == TFW_HTTP_HDR_RAW
-		     && hm->h_tbl->off == hm->h_tbl->size))
+	if (unlikely(id == TFW_HTTP_HDR_RAW))
 	{
```
Comment (krizhanovsky) [inline tempesta_fw/http_parser.c]: Place braces at the same line with IF statement if the statement is single-line.

File: tempesta_fw/http_parser.c [hunk 3]
```diff
@@ -864,64 +874,47 @@ __parse_transfer_encoding(TfwHttpMsg *msg, unsigned char *data, size_t *lenrval)
...
 		/* Allocate some more room if not enough to store the header. */
-		size_t order = hm->h_tbl->size / TFW_HTTP_HDR_NUM;
-
-		if (unlikely(__HHTBL_SZ(order + 1) >= TFW_HTTP_HDR_NUM_MAX)) {
```
Comment (krizhanovsky) [inline tempesta_fw/http_parser.c]: The line exceeds 80 characters

**Comment by krizhanovsky** [comment]
Unit and functional tests are definitely very wished things!

File: tempesta_fw/str.c [hunk 4]
```diff
@@ -80,61 +84,153 @@ validate_key(const char *key, int len)
...
+
+	IF_DEBUG {
+		validate_tfw_str(str);
+		BUG_ON(!start_pos);
+		BUG_ON(!end_pos);
+		BUG_ON(!pool);
+		BUG_ON(str->cnum == TFW_STR_CNUM_MAX);
+		BUG_ON((end_pos - start_pos) > TFW_STR_CHUNK_LEN_MAX);
+		BUG_ON(((end_pos - start_pos) + str->len ) > TFW_STR_LEN_MAX);
+		BUG_ON(end_pos < start_pos);
```
Comment (krizhanovsky) [inline tempesta_fw/str.c]: BUG_ON() is no-op statemen w/o DEBUG compile option (I mean kernel's, not Tempesta), so there is no need to place BUG_ON() statements into IF_DEBUG. validate_tfw_str() also must be empy statement w/o DEBUG compilation flag.

**Comment by vdmit11** [comment]
The TfwStrChunk idea was recognized useless during the review.
Therefore, the commit 1764413 and further changes are archieved and not merged to the master branch.
Previous commits are merged to the master branch manually.

**Pull request closed without merging**